[CK_TILE] Support WMMA (gfx12) in FMHA

ex-rzr opened the pull request:
## Proposed changes

Changes worth mentioning:
* The split KV combine kernel is more universal now, it supports any power-of-2 number of warps, any number of splits and kN1 = 16, see https://github.com/ROCm/composable_kernel/pull/2528/commits/e08a84d56d7babdccb4d42e9b3b7262c45678b11
* Some of hardcoded values (M0 of the dot bwd kernel, N1 of the split KV combine kernel) are set based on tile size, N1 of the split KV combine kernel is ensured to be consistent with the split KV kernel.
* Instances (tile size + pipeline) are checked for duplications and proper ordering of paddings (from non-padded to fully padded).
* A workaround for a compiler bug (SWDEV-559729) in bwd kernels with padding.
* WarpGemmDispatcher is used when possible instead of a concrete warp gemm implementation.
* Some tile distributions are updated to support different tile sizes and warp size.
* Initially this PR included a big change for supporting 16x16 warp tiles in `BlockDropout`, they were merged to develop as a part of #2836.

The main change is about supporting different architectures in one binary (for example, tests and examples) without a need to compile separate binaries for WMMA+warpsize32 and MFMA+warpsize64.

I extended the `KernelComponentFactory` mechanism, already existing in fwd, now all APIs work in this way:
* There are different factories for gfx12 and gfx9, with different tile sizes and pipeline parameters.
* There is also a factory for gfx950 based on gfx9 which adds pipelines with transpose loading (where suitable).
* Later additional factories can be added (for example, for gfx1250).
* Cmake scripts pass the list of targets to python scripts that generate FMHA kernels.
* Each kernel file is supposed to be built for one architecture "family" (gfx12, gfx9, gfx950), file names include corresponding suffixes.
* Building for other architectures is disabled by a special preprocessor guard.
* The `fmha_*_api.cpp` files check what the current device is and launches the required kernel, if there is no kernel for the current device (only for other architectures), -1 is returned, aka "not supported yet" instead of a crash with `hipErrorNoBinaryForGpu`.
* Each kernel includes a special tag for architecture, I added `gfx9_t` and `gfx950_t` to existing `gfx11_t` and `gfx12_t`. This is needed to link two kernels with the same template parameters built for different architectures (otherwise the linker keeps only one of the symbols with the same symbol name).
* Each host dispatching function (like `fmha_fwd_`) has a new template parameter with the architecture tag for a similar reason: having two host functions with the same traits built in files for different architectures leads to duplication symbols. These tags are optional and, in my opinion, are backward compatible.

`fmha_fwd_d..._gfx12.cpp`:
```cpp
...
#if !defined(__HIP_DEVICE_COMPILE__) || defined(__gfx12__)

using fmha_arch_tag = ck_tile::gfx12_t;
...
template<>
float fmha_fwd_<trait_0, fmha_arch_tag>(const ck_tile::stream_config& s, fmha_fwd_args a)
{
...
}
#endif // !defined(__HIP_DEVICE_COMPILE__) || defined(__gfx12__)
```

`fmha_fwd_api.cpp`:
```cpp
float fmha_fwd(fmha_fwd_traits t, fmha_fwd_args a, const ck_tile::stream_config& s) {
...
    [[maybe_unused]] const std::string device_name = ck_tile::get_device_name();

    if(device_name.compare(0, 6, "gfx950") == 0) {
        if(t.data_type.compare("fp16") == 0) {
            if(t.hdim_q <= 32 && t.hdim_v <= 32) {
                ...
                    return fmha_fwd_<trait_, ck_tile::gfx950_t>(s, a);
                ...
    }
    else if(device_name.compare(0, 5, "gfx12") == 0) {
        ...
    else if(device_name.compare(0, 4, "gfx9") == 0) {
        ...
    }
}
```

Here are examples of the generated files:
* [fmha_fwd_d256_bf16_group_b64x64x32x256x32x256_r4x1x1_r4x1x1_w16x16x16_w16x16x16_qr_vr_psskddv_nlogits_alibi_nmask_nlse_ndropout_nskip_nsquant_ntrload_gfx12.cpp](https://github.com/user-attachments/files/22883415/fmha_fwd_d256_bf16_group_b64x64x32x256x32x256_r4x1x1_r4x1x1_w16x16x16_w16x16x16_qr_vr_psskddv_nlogits_alibi_nmask_nlse_ndropout_nskip_nsquant_ntrload_gfx12.cpp)
* [fmha_fwd_api.cpp](https://github.com/user-attachments/files/22883426/fmha_fwd_api.cpp)

Please, let me know if something in this description is not clear.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

ex-rzr committed:
Pass hdim to tile_example_fmha_fwd in fp8 tests

ex-rzr committed:
Add WMMA support to fwd FMHA pipelines

ex-rzr committed:
Tune tile sizes a bit for less spilling

fp16 256 is still quite slow

ex-rzr committed:
Support 16x16 (MFMA, WMMA) and 32x32 (MFMA) tiles in fwd and bwd BlockDropout

Add comments with dropout implementation details

Fix performance regression of fwd+dropout

    * Remove some usage of type punning (reinterpret_cast with ref or ptr) in Philox;
    * "scalarize" seed and offset, they may come either from kernel args or from device memory
      (presumably loaded with vector loads).

    These changes help the compiler to procude more optimal code and reduce register spilling.

Use WarpGemmDispatcher instead of explicit WarpGemmMfma... to get  CWarpDstrEncoding

Use code based on BlockDropout in BlockDropoutBwd

Refactor BlockDropout (fwd)

Implement BlockDropout (fwd) for WMMA

    Originally BlockDropout only supported 32x32 tiles (IsWG32 = true),
    this version supports 16x16 tiles.
    If MPerBlock > MWarp * 16, it can generate numbers for two 16x16 tiles, similarly
    to BlockDropoutBwd.

Implement BlockDropoutBwd for WMMA

Remove MakeRandValLds* functions unused in BlockDropoutBwd

Remove unused Run overload from BlockDropoutBwd

ex-rzr committed:
Fix Q grad tile distribution for warp size = 32 and hdim >= 256

With AccDataType = float and warp size = 32, K0 becomes 0, K repeat is required to correcty distribute the tile.

ex-rzr committed:
Use code based on BlockDropout in BlockDropoutBwd

ex-rzr committed:
Fix split KV combine kernel for gfx12 (warp size 32) and make it more universal

* Fix LSE LDS tensor descriptors: kMaxSplits and kM0 were swapped, it worked on gfx9
  because they both equal to 8 while on gfx12 they are 8 and 4;
* Fix Oacc LDS tensor descriptor: it was transposed even though its shape=[4 * kM0, kN1],
  it worked on gfx9 because 4 * kM == kN1 == 32;
* Removing these hidden dependecies allows to support:
    * any number of warps (power-of-2), not only 4;
    * kN1 = 16, not only 32;
    * any number of splits;

ex-rzr committed:
Rename ids like o_acc_4 and Oacc4 to eliminate confusion: kNumWarps doesn't have to be 4 now

ex-rzr committed:
Replace hard-coded kN1 in dispatch code with the requested tile size

ex-rzr committed:
Add gfx12-specific tile sizes for split KV

ex-rzr committed:
Pass GPU architecture to kernel generation scripts

This is still a temporary solution.

ex-rzr committed:
Build and run FMHA CI tests for gfx12

ex-rzr committed:
Fix issue after merging

ex-rzr committed:
Fix bwd tile sizes

The current pipelines always read only one tile K and V tile, this
requires bk0 == bhdq and bk2 == bhdv (kK0 == kQKHeaddim and
kK2 == kVHeaddim).

ex-rzr committed:
Use hardware f32->f8 on gfx12, remove v_perm

__builtin_amdgcn_perm is not needed because
__builtin_amdgcn_cvt_pk_fp8_f32 allows to specify which word (16 bit of
 32-bit dword) is used to store results (two f8 values).

ex-rzr committed:
Update changelog

ex-rzr committed:
Add WMMA support to pagedkv

ex-rzr committed:
Fix regression with philox seed and offset when they exceed 32-bit int

__builtin_amdgcn_readfirstlane works with 32-bit values, seed and offset
are 64-bit so they get truncated.

ex-rzr committed:
Fix selection of a fallback tile based on bm0

The assumption that the largest bm0 == 128 is not always true for
current fp32 tiles.

ex-rzr committed:
Fix scripts after rebasing

ex-rzr committed:
Fix names after cherry-picking

ex-rzr committed:
Do not use filters related to qr_async_trload

They disable tiles/pipelines which are valid for gfx12.

ex-rzr committed:
Use different dstr encoding when C is transposed

ex-rzr committed:
Do not call GetQKBlockGemm (and hence WarpGemmDispatcher) in host code

Some WarpGemmDispatcher instantiations are defined only
for specific archs and undefined on host.
Calculations related to sched barriers are moved from Pipeline's public
fields into pipeline's operator().

ex-rzr committed:
Fix incorrect name WarpGemmMfmaFp8Fp8F32M32N32K16SwizzleBTransposedCDistribution

Correct name is WarpGemmMfmaFp8Fp8F32M32N32K32SwizzleBTransposedCDistribution
because it's 32x32x16 with IterateK = 2 so K = 32, also all tiles used
in codegen scripts are 32, 32, 32.

ex-rzr committed:
Generalize usages of WarpGemmDispatcher for MFMA and WMMA

WarpGemmMfmaFp8Fp8F32M32N32K32SwizzleBTransposedCDistribution is still
used explicitly becaus of swizzle factor = 4.

ex-rzr committed:
Mark has_load_tr as maybe_unused

There are no transpose loading for RDNA.

ex-rzr committed:
Remove CK_TILE_USE_MFMA/WMMA from fmha-related code

ex-rzr committed:
Detect BlockSize on host based on warp size of the current device

If kBlockSize == kNumWarps * get_warp_size(), the kernel is launched with
kBlockSize / 2 because on host get_warp_size() == 64 always.

ex-rzr committed:
Fix calculation of grid size for combine kernel with warp size = 32

ex-rzr committed:
Add missing includes and header

ex-rzr committed:
Support multiple archs in one binary for fwd

ex-rzr committed:
Support multiple archs in one binary for fwd_splitkv, fwd_appendkv, pagedkv_prefill

ex-rzr committed:
Support multiple archs in one binary for bwd

* trload kernels are compiled only for gfx950;
* instances with padding are checked after instances without padding so
  they can be used as fallbacks (similarly to fwd);

ex-rzr committed:
Extract common code from register_traits

ex-rzr committed:
Revert "Fix regression with philox seed and offset when they exceed 32-bit int"

To simplify merging , the proper fix is in develop already.

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Support new numerical d paddings in trait ordering checks

ex-rzr committed:
Build fp32 tests only on gfx9

ex-rzr committed:
Do not use hardcoded M0 = 64 for dot bwd kernel

ex-rzr committed:
Use textwrap.indent from standard library

ex-rzr committed:
Make fp8 pipelines on gfx12 consistent with gfx9

ex-rzr committed:
Update tests for current pipelines

ex-rzr committed:
Make ninja check more responsive in CI

ninja buffers output so this job looks hanging.

ex-rzr committed:
Support fp8fp32 by limiting O vector size

The fp32 output type requires storing 8 * sizeof(float) = 32 bytes,
which is not implemented (here 8 is the number of C values per lane for
v_wmma_f32_16x16x16...).

ex-rzr committed:
Remove unused cmake options

ex-rzr committed:
Unify including  amd_buffer_addressing.hpp/_builtins.hpp

ex-rzr committed:
Temporarily use amd_buffer_addressing.hpp on >=gfx10

amd_buffer_addressing_builtins.hpp uses inline asm for loads/stores
which is not compatible with >=gfx10:
 * 1 scalar for exec masks instead of 2,
 * gfx12 uses different instruction names etc.

ex-rzr committed:
Update asm in bf16 conversions to work with warp 32

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Do not generate splitkv/appendkv with vlayout=col for consistency with fwd

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Add arch tags to kernels/host funcs, compile for each arch separately

ex-rzr committed:
Add kM0 to fmha_bwd_dot_do_o kernel name to match filename

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Add workaround for miscompilation of bwd with padded hdim

SWDEV-559729: v_wmma instructions can be incorrectly placed in divergent
branches used to store padded tensors (when some lanes are inactive due
to padding). Inline asm with dummy dependencies on VGPRs of the tensors
prevents the compiler doing this.

ex-rzr committed:
Fix add_gtest_executable for absolute paths

Some tests (like gemm_tile_engine) pass absolute paths to source files.
In CI the branch name is a part of the root dir, and if the branch name
contains "wmma", "xdl" etc., files can be incorrectly excluded.

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Run only hdim 128 smoke tests for fp8fp32

There are no instances for hdim 64 and 256.

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Merge branch 'develop'

ex-rzr commented:
I have added a description of the changes. Ready for review

ex-rzr committed:
Format py with ruff to simplify merging develop

ex-rzr committed:
Merge branch 'develop'

illsilin commented on include/ck_tile/core/tensor/buffer_view.hpp:
I don't quite understand this part.
The purpose of amd_buffer_addressing_builtins.hpp was to add all the new built-ins supported by newer compiler versions, which will never be available in the older compilers. Hence the #if/#else switch based on the compiler version. Otherwise they should offer identical functionality. So what's the reason to include both of them?

ex-rzr replied:
The problem is that, despite the name, `amd_buffer_addressing_builtins.hpp` doesn't use builtins, only inline assembly which is not compatible with RDNA4 (different instruction names for loads/stores, 32-bit exec masks etc.).

I don't understand why it's done this way.
And I don't have enough information to remove `amd_buffer_addressing_builtins.hpp`, because it's possible that inline assembly versions are faster in some cases (especially finctions with conditional loads/stores that explicitly set and restore exec mask).

It seems that this is a topic for a separate investigation as it requires performance evaluation of both solutions in real applications. 

In this file I just removed the check of the compiler version to make it consistent with two other places (`core.hpp` and `arch.hpp`) where they are included like this:
```
#include "ck_tile/core/arch/amd_buffer_addressing_builtins.hpp"
#include "ck_tile/core/arch/amd_buffer_addressing.hpp"
```
`amd_buffer_addressing.hpp` and `amd_buffer_addressing_builtins.hpp` have guards with `CK_TILE_USE_BUFFER_ADDRESSING_BUILTIN`, this definintion is set in `include/ck_tile/core/config.hpp`.

So overall I didn't change functionality, just forced to use the version with builtins on >=gfx10 (in `config.hpp`) and removed `__clang_major__ >= 20` as its already controlled by `CK_TILE_USE_BUFFER_ADDRESSING_BUILTIN`.

illsilin replied:
OK, we'll take care of this in a separate PR.

linqun commented on include/ck_tile/host/kernel_launch.hpp:
```diff
@@ -28,18 +28,47 @@ __launch_bounds__(Kernel::kBlockSize, MinBlockPerCu)
 // the "static __device__ operator()(some_arg)" is the entry point of KernelImpl
 //
-template <int MinBlockPerCu = CK_TILE_MIN_BLOCK_PER_CU, typename KernelImpl, typename... Args>
+// Arch can be used to support linking multiple object files that have the same kernel compiled for
```
what will happen if we don't add arch in kenerl template? link error or something else?
I think compiler should have handled kernel on multi device.

ex-rzr replied:
Let's assume that we generate the same kernel (same tile size and pipeline) for different architectures, gfx9 and gfx12. Each is built as a separate file: `fmha..._gfx9.o` and `fmha..._gfx12.o`. So these two files use the same symbol for the kernel built for different architectures, the linker does not combine them into one "fat" symbol but keeps only one of them in the resulting binary.

In this case we can get a **runtime** error that there is not kernel for the current device (and on another device it works).

linqun reviewed (commented):
looks good to me for others.

joyeamd commented on example/ck_tile/01_fmha/script/smoke_test_fwd.sh:
```diff
@@ -114,9 +114,9 @@ run_fp8fp32_tests() {
     for bias in "n" "e" "a" ; do
     for b in 1 2 ; do
-    for hdim in 64 128 256 ; do
+    for hdim in 128 ; do
```
him 64, 256 is removed; need to restore.

ex-rzr replied:
But there are no instances for hdim != 128: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py#L672

Previously it didn't matter because it had `-d=128` and ignored the actual `hdim`

ex-rzr committed:
Merge branch 'develop'

joyeamd commented on test/ck_tile/fmha/test_fmha_fwd.cpp:
```diff
@@ -38,7 +38,7 @@ struct TestConfigs
         std::tuple{32, -1}, std::tuple{64, -1}, std::tuple{128, -1}, std::tuple{256, -1}};
     static constexpr auto ModeValues        = std::array{mode_enum::batch, mode_enum::group};
-    static constexpr auto IsVRowmajorValues = std::array{false, true};
+    static constexpr auto IsVRowmajorValues = std::array{true};
```
Maybe also need to restore back IsVRowmajorValues = std::array{false, true}

ex-rzr replied:
A similar reason, currently there are no instances with `"col"`: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py#L699

Of course, in this case the test cases with IsVRowmajorValues = false would be marked as SKIPPED, not a big problem

ex-rzr commented:
It seems that my changes in the codegen scripts are not compatible with aiter and pytorch.
I'm working on this. Perhaps passing `--targets` will be enough, I need to check it.

ex-rzr committed:
Fix incorrect var name

ex-rzr committed:
Codegen for gfx9,gfx950 when --targets is not specified

Aiter and Pytorch require changes for passing their targets to the codegen scripts.
With this temporary solution the files are generated but not all of them
have to be really built (depending on the used --offload-arch=).

ex-rzr committed:
Combine arch-related values into ArchTrait

This more centralized approach removes duplication of various formatting templates.

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Try a workaround for Jenkins error "groovyjarjarasm.asm.MethodTooLargeException: Method too large"

Some code is extracted into a function.

ex-rzr commented:
One of the new commit is mostly refactoring: less error-prone duplication of arch-related formatting templates in every codegen script.

Another moves some code from Jenkins pipeline to a separate function to work around the Jenkins crash with "Method too large" (CI jobs couldn't even started). Apparently merging the recent develop branch with some additions to the pipeline plus my changes in this branch made the pipeline too large to be compiled as a single Java method. My fix works but the issue may return later (There is a JVM property  recommended in https://issues.jenkins.io/browse/JENKINS-37984 as a workaround for a similar issue, I don't know if it helps with this one).

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Merge branch 'develop'

ex-rzr committed:
Merge branch 'develop'
